### ISSUE TITLE
Consolidation ttl: `spec.disruption.consolidateAfter`

### ISSUE THREAD

**Issue by runningman84** [kubernetes-sigs/karpenter#735]
### Tell us about your request

We have a cluster where there are a lot of cron jobs which run every 5 minutes...

This means we have 5 nodes for our base workloads and every 5 minutes we get additional nodes for 2-3 minutes which are scaled down or consolidated with existing nodes.

This leads to a constant flow of nodes joining and leaving the cluster. It looks like the docker image pull and node initialization creates more network traffic fees than the cost reduction of not having running the instances all the time.

It would be great if we could configure some time consolidation period maybe together with ttlSecondsAfterEmpty which would only cleanup or consolidate nodes if the capacity was idling for x amount of time.

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

Creating a special provisioner is quite time consuming because all app deployments have to be changed to leverage it...

### Are you currently working around this issue?

We think about putting cronjobs into a special provisioner which would not use consolidation but the ttlSecondsAfterEmpty feature.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by thelabdude** [comment]
The documentation in 0.32.1 for `consolidateAfter` is very confusing. Specifically, it states:
```
ConsolidateAfter is the duration the controller will wait before attempting to terminate nodes that are underutilized.
```
But it's not compatible with `WhenUnderutilized`! From the description, it sounds like it applies to "nodes that are underutilized" but is not the case.

**Comment by mullermateus** [comment]
Hello everyone! I do agree with @thelabdude. I was expecting to use `consolidateAfter` along with `WhenUnderutilized`.

Our use case is similar to the ones mentioned above. I would really like Karpenter to optimize underutilized workloads.

The problem is that this can't happen during our application deployment, otherwise Spinnaker gets lost during enable/disable traffic for the blue/green deployment. Spinnaker will try to add a label to a pod that doesn't exist anymore due to Karpenter deleting the nodes.

The `consolidateAfter` works perfectly for this, but should be along with `WhenUnderutilized`.

I hope that make sense. Do we have any plans to support this?

**Comment by sftim** [comment]
If you want to protect Pods during deployment, you could temporarily annotate them as exempt from consolidation @mullermateus. In other words, existing features let you consolidate underused nodes, and to protect Pods from eviction linked to that consolidation.

**Comment by sftim** [comment]
https://github.com/kubernetes-sigs/karpenter/issues/696 also proposes a mechanism to delay the scale-in.

**Comment by edas-smith** [comment]
Definitely a feature that would be good to have, is there any further update on this/whether its being worked on?

**Comment by Hronom** [comment]
Hello, is there any news?

I would heavy need this.

Expected to make `consolidateAfter` works with `WhenUnderutilized`.
Like described here https://github.com/kubernetes-sigs/karpenter/issues/735#issuecomment-1864822421

**Comment by almson** [comment]
> The only way to get the semantic to be technically correct is to recompute the consolidatability for the entire cluster on every single pod creation/deletion. The algorithm described above is a computationally feasible way (equivalent to current calculations), but has weird edge cases. Would you be willing to accept those tradeoffs?

You are overthinking this!

Just help us reduce node thrashing FFS.

**Comment by Luke-Smartnews** [comment]
FYI: https://github.com/kubernetes-sigs/karpenter/pull/992

**Comment by almson** [comment]
> The only way to get the semantic to be technically correct is to recompute the consolidatability for the entire cluster on every single pod creation/deletion.

@ellistarn I disagree. The following algorithm might be "correct" and effective:

1) Record the current state of the cluster
2) Run a simulation of the cluster and record its proposed consolidated state. (expensive)
3) On every pod creation, try to schedule the pod on both the current and consolidated cluster (cheap)
4) If the pod schedules on the current but not the consolidated cluster, then that would indicate thrashing.
  4b) Re-run the simulation at this "high-watermark" to create a new proposed less-consolidated state.
  4c) Reset timer
  4d) go to step 3
5) If 'consolidateAfter" elapses without running into situation 4, then proceed with the consolidation.

**Comment by sftim** [comment]
I've advocated for some more nuanced mechanisms (in other words, rejecting a plain TTL approach). If we only allow the simple thing, we might make it harder to support more complex mechanisms. I'm keen that if we do paint ourselves into an API design corner, we do it with our eyes open.

If the simple TTL approach is what folks are keen to have, I wonder if we can allow either:
`.spec.consolidationMinimumDelay` (day 1)
`.spec.consolidationPolicy` (day _n_, rules specified in CEL based on values exposed by Karpenter; potentially, reevaluated whenever one input value changes)

**Comment by Hronom** [comment]
@almson , @sftim hey please, for beginning just add delay before nodes get consolidated like described in comment https://github.com/kubernetes-sigs/karpenter/issues/735#issuecomment-1864822421

You can think about other non trivial mechanisms later on. 
For now it will be enough.
Also it okey if you change API later on, you already did like this with previous Karpenter releases, I believe nobody will blame you.

You did awesome product, please add this and I think it will transform from awesome to perfect=)

**Comment by victorboissiere** [comment]
We have the exact same issue. Can't we learn from cluster autoscaler and implement the feature scale-down-delay? (which would look like `.spec.consolidationMinimumDelay`). It sounds like we lost a need covered in the past by moving to Karpenter. We have hundreds of developers complaining of pods being disrupted due to scaling up/down events, some pods are disrupted more than 20 times in an hour. Having a scale-down delay will mitigate this.
I am considering reverting the move and switching back to cluster autoscaler just for that missing feature.
Let me know if we can help on this & how (design document, implementation, etc.).

**Comment by sidewinder12s** [comment]
Another complication that Karpenter makes worse is that certain metrics vendors may bill based on unique metrics series produced over a month. Karpenter's consolidation behavior will actively make this worse as it replaces/reschedules workloads and it'd be useful if we could constrain how often this might happen.

**Comment by rtomadpg** [comment]
I agree default scaledown / consolidation behaviour is too aggressive in some use cases.

Most Karpenter users have previously used Cluster Auto Scaler which is much much slower in scale downs. From CAS docs at https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-down-work:

> If a node is unneeded for more than 10 minutes, it will be terminated [by Cluster Auto Scaler]

I understand Karpenter's key purpose is to avoid unnecessary cloud cost, but in some cases Karpenter operators disagree with what Karpenter deems unnecessary. For CI/CD and other Jobs heavy use cases, having young nodes stick around while underutilized, because new work is expected, is great for devex and justifies cost. IMHO.

Something like the suggested `.spec.consolidationMinimumDelay` would meet the basic need of delaying consolidation. Other advanced methods would probably be used less?

**Comment by sftim** [comment]
> For CI/CD and other Jobs heavy use cases, having young nodes stick around while underutilized, because new work is expected, is great for devex and justifies cost. 

I (still) think we're missing an opportunity to use placeholder Pods to provide overhead capacity, leaving the scheduler to preempt those placeholders as appropriate and for Karpenter to stick to its knitting around automatic scaling.
A controller to manage those placeholder Pods could make a great addition to the Kubernetes project.
However, we're missing a really good set of user stories that we can use to try to fit solutions against.

(Would anyone like to work on collating that set of user stories?)

**Comment by sftim** [comment]
One thing to remember: if we add `.spec.consolidationMinimumDelay`, we probably have to live with that decision for a decade or more. Putting things into an API that will graduate to stable has costs as well as benefits.

**Comment by rtomadpg** [comment]
Hi @sftim 

> I (still) think we're missing an opportunity to use placeholder Pods to provide overhead capacity, leaving the scheduler to preempt those placeholders as appropriate and for Karpenter to stick to its knitting around automatic scaling.
> A controller to manage those placeholder Pods could make a great addition to the Kubernetes project.

This feature already exists and does not even need a (new) controller. A Deployment managing N right-sized low priority pause 'placeholder' pods, should keep the nodes busy. However, you probably want to add scheduled scale up/downs, to avoid having these placeholder pods run 24/7 (and rack cost). That's DIY. For the basics, have a look at https://github.com/codecentric/cluster-overprovisioner/ (not my project).

> Putting things into an API that will graduate to stable has costs as well as benefits.

I totally agree. However, I believe the Karpenter people have ample experience adding features and later on extending/replacing them, so let's see what the future will bring.

**Comment by sftim** [comment]
> Why not use the v0 version

Ah. We're at beta (API version v1beta1) and are [planning](https://github.com/kubernetes-sigs/karpenter/issues/758) to stabilize the API (v1).

Kubernetes has a particular approach to API version round trips, that we intend to follow. So adding things in even at beta has constraints.

**Comment by Hronom** [comment]
@sftim there no need to change API, here is the direct and clear proposal: 
Use property that [karpenter already have](https://karpenter.sh/docs/concepts/nodepools/) `consolidateAfter` and make it work with `consolidationPolicy` being set `WhenUnderutilized`. It's fits ok in API from architectural perspective.

## What is needed
### Item 1
Implement logic that will enable users to use `consolidateAfter`  when `consolidationPolicy` being set to `WhenUnderutilized`.

Behavior of this logic will be same as behavior when `consolidationPolicy` set to `WhenEmpty`.
When `consolidationPolicy` set to `WhenEmpty` and for example we set `consolidateAfter` to `30s` it will delay consolidations for `30s`.

### Item 2
Adjust comment [here in documentation](https://karpenter.sh/docs/concepts/nodepools/).
Change from:
```
...
  disruption:
    # Describes which types of Nodes Karpenter should consider for consolidation
    # If using 'WhenUnderutilized', Karpenter will consider all nodes for consolidation and attempt to remove or replace Nodes when it discovers that the Node is underutilized and could be changed to reduce cost
    # If using `WhenEmpty`, Karpenter will only consider nodes for consolidation that contain no workload pods
    consolidationPolicy: WhenUnderutilized | WhenEmpty

    # The amount of time Karpenter should wait after discovering a consolidation decision
    # This value can currently only be set when the consolidationPolicy is 'WhenEmpty'
    # You can choose to disable consolidation entirely by setting the string value 'Never' here
    consolidateAfter: 30s
...
```
to
```
...
  disruption:
    # Describes which types of Nodes Karpenter should consider for consolidation
    # If using 'WhenUnderutilized', Karpenter will consider all nodes for consolidation and attempt to remove or replace Nodes when it discovers that the Node is underutilized and could be changed to reduce cost
    # If using `WhenEmpty`, Karpenter will only consider nodes for consolidation that contain no workload pods
    consolidationPolicy: WhenUnderutilized | WhenEmpty

    # The amount of time Karpenter should wait after discovering a consolidation decision
    # You can choose to disable consolidation entirely by setting the string value 'Never' here
    consolidateAfter: 30s
...
```

**Comment by qibint** [comment]
After upgrading to v0.32.8 from v0.31.0, I see more aggressive node launching and termination operations with the same request pattern. I followed the https://karpenter.sh/v0.32/upgrading/v1beta1-migration/#ttlsecondsafterempty to switch `ttlSecondsAfterEmpty` of v1alpha to the new 
```
consolidationPolicy: WhenEmpty
consolidateAfter: 2m
```
of v1beta1, however, no matter how I set up the consolidateAfter event to `60m`, the node gets terminated quickly which makes the running node reusable impossible.

Created an issue here - https://github.com/aws/karpenter-provider-aws/issues/5938.
Starting March 26th, there are an increasing number of pending pods. This upgrade makes the scheduler less responsive due to the unnecessary nodes registering and leaving the cluster.

<img width="813" alt="Screenshot 2024-03-27 at 10 45 41 AM" src="https://github.com/kubernetes-sigs/karpenter/assets/41122460/cabee636-924c-4035-a646-0013f88ff99f">

Could anyone have a sugesstion regarding this problem? The configuration and logs are in the above issue link.

Thanks

**Comment by vainkop** [comment]
+1 to make `consolidateAfter` work with `WhenUnderutilized`!
Thanks @Luke-Smartnews for your #992, I hope it's merged soon!

**Comment by morremeyer** [comment]
The state of this issue is confusing:

* Specifying `consolidateAfter` together with `consolidationPolicy: WhenUnderutilized` does lead to an error: `spec.disruption: Invalid value: "object": consolidateAfter cannot be combined with consolidationPolicy=WhenUnderutilized`
* The [Design Document](https://github.com/kubernetes-sigs/karpenter/blob/43da3604c14097ab1ee3ca360af5a82446a4db4d/designs/disruption-controls.md) makes no mention of this which leads me to expect that this should be supported since the [documentation for v0.35](https://karpenter.sh/v0.35/concepts/disruption/#consolidation), which we run, does not mention that it is not possible
* The documentation at [karpenter.sh](https://karpenter.sh) links to the [aws/karpenter-provider-aws](https://github.com/aws/karpenter-provider-aws) repository, not to this one. 
* The [migration documentation](https://karpenter.sh/v0.32/upgrading/v1beta1-migration/#consolidation) mentions `Note: You currently can’t set the consolidateAfter field when specifying consolidationPolicy: WhenUnderutilized. Karpenter will use a 15s consolidateAfter runtime default.`, a hint that is not present in the v0.35 documentation. This again was leading me to believe that it is supported in v0.35. I have opened https://github.com/aws/karpenter-provider-aws/pull/6016 to add this hint.
* #992 has been closed with the note that an RFC needs to be created

It would be very helpful if someone could clarify:

* Is anybody working on an RFC yet? If not, what needs to be done to open an RFC? I couldn't find any mention of this in the documentation in this repository except for in https://github.com/kubernetes-sigs/karpenter/blob/43da3604c14097ab1ee3ca360af5a82446a4db4d/designs/nodepool-requirement-flexibility.md?plain=1#L13
* Is the documentation at karpenter.sh in the https://github.com/aws/karpenter-provider-aws repository for historical reasons? Is that the official documentation for karpenter or only for karpenter running on AWS EKS?

Thanks y'all!

**Comment by alicancakil** [comment]
I need this so badly to reduce churn in my cluster.

**Comment by sftim** [comment]
@alicancakil are you hoping to help with code changes in this area? If so, we may be able to offer some guidance.

**Comment by joewragg** [comment]
+1

**Comment by sansice** [comment]
Can something like **consolidateNotBefore** help us and easy to implement? - [The node can be a candidate for consolidation only after certain time the last pod has been created in the node]

**Comment by sansice** [comment]
Can something like this help us and easy to implement? - [The node can be a candidate for consolidation only after certain time the last pod has been created in the node]

**Comment by morremeyer** [comment]
@njtran, you are the assignee here. Could you maybe clarify the questions in https://github.com/kubernetes-sigs/karpenter/issues/735#issuecomment-2047747008 or provide an outlook of how we can proceed here?

**Comment by Nuru** [comment]
@ellistarn @njtran How about this for an implementation algorithm:

- When a node is seen to be underutilized, it is annotated with a taint `consolidationCandiate:true:NoSchedule` and possibly with a label giving the time it can be consolidated (now + `consolidateAfter`)
- When a new Pod is deployed, Karpenter tries to fit it into an untainted Node. If it cannot, before scaling up, it considers Nodes with the `consolidationCandidate` taint. If the Pod will fit on one of those nodes, and after doing so, cause the Node to no longer be UnderUtilized, the taint gets removed and the Pod gets placed on that node. (If there is a batch, this can be applied to all or a portion of the batch.) If there is no such tainted node, then a new node is launched.
- Later, when a consolidation pass is run by the disruption controller, any nodes whose `consolidateAfter` duration have passed and are still tainted/underutilized, get irreversibly marked for termination. Displaced pods are candidates for being placed on other underutilized nodes. 

In other words, underutilized Nodes are treated as something like a "Warm Pool", and terminated if they remain in the warm pool longer than `consolidateAfter`. If terminating the underutilized Node would cause a replacement Node of the same size to be launched (due to other cluster requirements), then at the point where it would otherwise be terminated, it can have its `consolidateAfter` timer restarted. (Or something like that. However Karpenter currently handles the spillover from bin packing that causes the cheapest available Node to be needed even though it is underutilized.)


Like other commentators, I have had to give up on `WhenUnderutilized` in favor of `WhenEmpty` in order to avoid the excessive churn with the immediate termination of underutilized nodes. This would fix that.

**Comment by booleanbetrayal** [comment]
> CAS has the `--scale-down-utilization-threshold` (along with the other flags mentioned) and seems like Karpenter needs a similar tunable. Unfortunately, we're seeing so much disruption in running pods b/c of consolidation that we can't use Karpenter in any of our active clusters.

I am also coming from a highly-tuned cluster-autoscaler configuration where `scale-down-utilization-threshold` seems to have provided for better bin-packing than Karpenter is currently providing. Is there any documentation for the thresholds that `WhenUnderutilized` implements? I haven't been able to find any specifics or hunt it down in code yet.

**Comment by Shadowssong** [comment]
I am also experiencing the high churn issue that has been reported here and until Karpenter supports more configurations around this we found a semi hacky middle ground in between always scaling up or constantly consolidating by using budget schedules:

```
  spec:
    disruption:
      budgets:
      - duration: 45m
        nodes: "0"
        schedule: '@hourly'
      - nodes: "1"
      consolidationPolicy: WhenUnderutilized
```

This basically limits consolidation to only happen during the last 15 minutes of each hour and only 1 node at a time.  This is by no means a good final solution but it seems to work as expected and you can fine tune the times to prevent constant consolidation.  Hope this helps anyone else looking for a temporary workaround.

**Comment by edas-smith** [comment]
Has there been any updates on when this may be implemented?

**Comment by vainkop** [comment]
> Has there been any update on when this may be implemented?

It seems the solution was already implemented in [this](https://github.com/kubernetes-sigs/karpenter/pull/1218) PR but it hasn't been merged yet :(

**Comment by edas-smith** [comment]
> > Has there been any update on when this may be implemented?
> 
> It seems the solution was already implemented in [this](https://github.com/kubernetes-sigs/karpenter/pull/1218) PR but it hasn't been merged yet 😞

Ah I see! Thanks for linking me to it :) glad this is just now pending a merge, which hopefully will come soon!

**Comment by njtran** [comment]
Hey all, after some more design discussion, we've decided not to go with the approach initially taken in 
https://github.com/kubernetes-sigs/karpenter/pull/1218, as it was computationally expensive, hard to reason about, and had to take too many guesses as to how the kube-scheduler would schedule pods in relation to our simulations.

Alternatively, the approach we're now running with for v1 is using `consolidateAfter` to filter out nodes for consolidation. When you set `consolidationPolicy: WhenUnderutilized` and (for example) set `consolidateAfter: 1m`, Karpenter will not consolidate nodes that have had a pod scheduled or removed in the last minute. To achieve the same Consolidation behavior  prior to v1, you can set `consolidateAfter: 0s`. 

Let us know what you think of this approach.

**Comment by vainkop** [comment]
> Hey all, after some more design discussion, we've decided not to go with the approach initially taken in #1218, as it was computationally expensive, hard to reason about, and had to take too many guesses as to how the kube-scheduler would schedule pods in relation to our simulations.
> 
> Alternatively, the approach we're now running with for v1 is using `consolidateAfter` to filter out nodes for consolidation. When you set `consolidationPolicy: WhenUnderutilized` and (for example) set `consolidateAfter: 1m`, Karpenter will not consolidate nodes that have had a pod scheduled or removed in the last minute. To achieve the same Consolidation behavior prior to v1, you can set `consolidateAfter: 0s`.
> 
> Let us know what you think of this approach.

@njtran please clarify: does it mean that starting with v1 `consolidateAfter` can be configured along with `consolidationPolicy: WhenUnderutilized` ? I believe an error or a warning is printed if v1beta1 is used.

Is there any eta for a Karpenter v1 API release with which the above config can actually be used?

Thank you.

**Comment by njtran** [comment]
@vainkop I mean when this is implemented, if you set `consolidateAfter: 0s`, it'll be backwards compatible, and present the same behavior as before this was implemented.

**Comment by wmgroot** [comment]
> Karpenter will not consolidate nodes that have had a pod scheduled or removed in the last minute. 

Are there any plans to support consolidation constraints that do not depend on pod scheduling at all, and instead depend on total node lifetime? I think relying solely on time since pod scheduling occurred could result in these issues, and may not meet our needs as cluster administrators.

1. Nodes may never be consolidated if pods are churning too frequently (due to frequent deploys or ephemeral jobs scheduling), resulting in inefficient cluster binpacking and wasted resources.
2. Nodes may be torn down before reaching a desired minimum age. We have license agreements with 3rd party providers such as Datadog that bill us for node with a 1 hour minimum. If a node exists for only 30m, we are still billed for the full hour. The implementation as described would require that we set our `consolidateAfter` to a minimum of 1 hour to avoid this issue, even if we would rather set it below an hour to better match our desired pod churn rates.

**Comment by wmgroot** [comment]
I do support and think having a consolidation control that's based on pod scheduling is a good idea, but I think it may fall short of meeting all use cases that are asking for better consolidation control. Perhaps having additional fields that limited consolidation would cover all use cases better. I think something like this is what we'd likely want to use if each of these fields were available.

```
consolidationOptions:
  minimumAge: 6h
  afterPodScheduledPeriod: 30m
  afterPodRemovedPeriod: 15m
```

**Comment by njtran** [comment]
> 1. Nodes may never be consolidated if pods are churning too frequently (due to frequent deploys or ephemeral jobs scheduling), resulting in inefficient cluster binpacking and wasted resources.
> 2. Nodes may be torn down before reaching a desired minimum age. We have license agreements with 3rd party providers such as Datadog that bill us for node with a 1 hour minimum. If a node exists for only 30m, we are still billed for the full hour. The implementation as described would require that we set our consolidateAfter to a minimum of 1 hour to avoid this issue, even if we would rather set it below an hour to better match our desired pod churn rates.

Totally agree here. I think a minimum age would be better solved through the `do-not-disrupt` annotation, supported on the node for your use-case, but also potentially the pod https://github.com/kubernetes-sigs/karpenter/issues/752. 

>  afterPodScheduledPeriod: 30m
>  afterPodRemovedPeriod: 15m

On these two configurations, I wonder if this would be better prescribed with some sort of node headroom, static capacity. This is really to combat the case where a set of job pods go down, and you want to reserve the capacity for some period of time because you know that pods will come back soon?

**Comment by ellistarn** [comment]
We've discussed a `minimumNodeLifetime`, that would apply to all disruption methods. I see this as orthogonal to this effort, and something we should explore in a followon.

**Comment by wmgroot** [comment]
> I think a minimum age would be better solved through the do-not-disrupt annotation

Are you suggesting that I wrote a controller to set the do-not-disrupt annotation based on node age myself, or that Karpenter would support this through use of the do-not-disrupt annotation under the hood as an implementation option?

>This is really to combat the case where a set of job pods go down, and you want to reserve the capacity for some period of time because you know that pods will come back soon?

`afterPodScheduledPeriod` would typically be to help with particular applications that have slow startup speeds (we have some stateful apps that rebalance data over several minutes on startup. I could also see it being a potentially better option than the do-not-disrupt annotation for ephemeral pods like CI jobs that typically run less than an hour but can lose a lot of progress or require manual retries if disrupted. 

`afterPodRemovedPeriod` is me taking a guess that a small buffer like this could be helpful with reducing node churn in the cluster, since pods could more easily schedule into gaps left by removed pods. This setting seems identical to CA's `scale-down-unneeded-time`
> Scaling down of unneeded nodes can be configured by setting --scale-down-unneeded-time. Increasing value will make nodes stay up longer, waiting for pods to be scheduled while decreasing value will make nodes be deleted sooner.
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-modify-cluster-autoscaler-reaction-time

> We've discussed a minimumNodeLifetime, that would apply to all disruption methods. I see this as orthogonal to this effort, and something we should explore in a followon.

Sure, but are we considering if I want to block consolidation disruption for nodes <6h old, but drift nodes immediately regardless of age? Implementing this as an isolated consolidation control may be the best way forward.

**Comment by vainkop** [comment]
> @vainkop I mean when this is implemented, if you set `consolidateAfter: 0s`, it'll be backwards compatible, and present the same behavior as before this was implemented.

I was of course more interested in the new functionality solving the issue we're discussing rather than smth that presents "the same behavior as before this was implemented"...

**Comment by njtran** [comment]
@wmgroot  

> Are you suggesting that I write a controller to set the do-not-disrupt annotation based on node age myself, or that Karpenter would support this through use of the do-not-disrupt annotation under the hood as an implementation option?

This would be a feature supported natively through Karpenter: linked the issue here https://github.com/kubernetes-sigs/karpenter/issues/752.

> afterPodRemovedPeriod is me taking a guess that a small buffer like this could be helpful with reducing node churn in the cluster, since pods could more easily schedule into gaps left by removed pods. This setting seems identical to CA's scale-down-unneeded-time

Makes sense here, we sort of follow this with our 15s validation period in between finding and executing consolidation. We could consider making this configurable in the future too.

**Comment by njtran** [comment]
> @njtran please clarify: does it mean that starting with v1 APIconsolidateAfter can be configured along with consolidationPolicy: WhenUnderutilized ? I believe an error or a warning is printed if v1beta1 is used.

@vainkop, yes you will be able to set `consolidateAfter` with `WhenUnderutilized`.

**Comment by tareks** [comment]
It may be worth going forward the proposed approach to see how it holds up in the real world.

At first glance, it seems like consolidation based on pod scheduling may not yield desired results, at least in our use case. The trigger to delete nodes appears to more of a "recentlyUtilized" condition (as opposed to `WhenUnderutilized`.

Here's one scenario:
1. Nodes are relatively bin packed. (initial state)
2. Multiple deployments kick off at the same time requiring more resources.
3. Karpenter provisions new nodes and pods are scheduled on them.
4. CronJob based pods start being scheduled on the new node(s) resulting in pods scheduled within a few minutes.
5. One or more of the new nodes may never be consolidated despite being under utilised because they pods constantly being scheduled on them.

I would expect that at some point the cluster will reach some state of equilibrium. Unclear if there may be runaway situations where one cold end up with several under utilised nodes.

Similar to what @Shadowssong mentioned, we've found that this sort of configuration may work as a tactical approach, especially in nonprod clusters:
```
    disruption:
      budgets:
      - nodes: 10%
      - nodes: "2"
      - duration: 8h
        nodes: "0"
        schedule: 0 9 * * mon-fri
      consolidationPolicy: WhenUnderutilized
      expireAfter: 720h0m0s
```

It reduces node churn by allowing a short window for karpenter to run it's consolidation cycle.  This, in turn, reduces the amount of disruption over the past week or so with (SPOT) nodes now showing significantly longer lifetimes/ages.

**Comment by njtran** [comment]
Another point worth noting for the approach mentioned in https://github.com/kubernetes-sigs/karpenter/issues/735#issuecomment-2214764162:

Karpenter prioritizes consolidating nodes that have the least amount of pods scheduled. Compare this to the default plug-in for kube-scheduler, which will schedule pods to the nodes that are the [least allocated](https://kubernetes.io/docs/reference/scheduling/config/#:~:text=NodeResourcesFit%3A%20Checks%20if%20the%20node%20has%20all%20the%20resources%20that%20the%20Pod%20is%20requesting.%20The%20score%20can%20use%20one%20of%20three%20strategies%3A%20LeastAllocated%20(default)%2C%20MostAllocated%20and%20RequestedToCapacityRatio.%20Extension%20points%3A%20preFilter%2C%20filter%2C%20score.). Generally, this means that Karpenter's heuristic is least compatible with the kube-scheduler default, scheduling pods to nodes that Karpenter wants to disrupt. If the kube-scheduler was configured to `MostAllocated`, Karpenter would run into less race cases, and be the most compatible with the proposed solution.

**Comment by wmgroot** [comment]
> It reduces node churn by allowing a short window for karpenter to run it's consolidation cycle. 

Implementing a maintenance window is undesired for us because we do not want to create special periods of time where disruption can occur vs not. This makes on-call awkward, increases the complexity of information we have to communicate to users of our clusters, and requires unique configuration of our NodePools for each region we operate in.

I'd also like to note that we need to be able to configure consolidation controls separate for consolidation types in some cases. We've currently implemented a patch to disable single-node consolidation altogether after finding it wasn't providing much value in return for the large amounts of disruption to our clusters. Given there's a tangible $$$ cost to restarting applications in some cases, it's entirely possible that single-node consolidation wastes more money than it saves with a naive implementation.

Since single-node consolidation exists solely to replace one EC2 instance with a cheaper variant, having a way to control the threshold at which karpenter will decide to replace the node would be wonderful (eg, only replace if you'll reduce the cost of the node by >15%).

**Comment by ellistarn** [comment]
> Since single-node consolidation exists solely to replace one EC2 instance with a cheaper variant, having a way to control the threshold at which karpenter will decide to replace the node would be wonderful (eg, only replace if you'll reduce the cost of the node by >15%).

+1 for price improvement threshold. I think this is orthogonal to consolidateAfter, though. Do you mind cutting a new issue for this, and referencing https://github.com/kubernetes-sigs/karpenter/blob/main/designs/spot-consolidation.md#2-price-improvement-factor?

**Comment by wmgroot** [comment]
Done.
https://github.com/kubernetes-sigs/karpenter/issues/1440

**Comment by miadabrin** [comment]
Is there a way to have this implemented/released sooner than v1? I think this will give everyone a way to handle the aggressiveness of the scaling down a little better

**Comment by samox73** [comment]
@miadabrin I think the urgency of this issue is evident from the lively discussion. Development for the issue is ongoing at #1218 and the discussion seems semi-active, but I'm unsure how close this PR is to completion.

**Comment by wmgroot** [comment]
@samox73 The maintainers have stated that the PR you linked has been abandoned in favor of a new implementation of consolidateAfter.

https://github.com/kubernetes-sigs/karpenter/issues/735#issuecomment-2214764162

**Comment by sherifabdlnaby** [comment]
> Is there a way to have this implemented/released sooner than v1? I think this will give everyone a way to handle the aggressiveness of the scaling down a little better

I second that, it won't be a breaking change per-se because it was disallowed to set `consolidateAfter` with `WhenUnderutilized`

**Comment by tm-nadavsh** [comment]
@Shadowssong 
> I am also experiencing the high churn issue that has been reported here and until Karpenter supports more configurations around this we found a semi hacky middle ground in between always scaling up or constantly consolidating by using budget schedules:
> 
> ```
>   spec:
>     disruption:
>       budgets:
>       - duration: 45m
>         nodes: "0"
>         schedule: '@hourly'
>       - nodes: "1"
>       consolidationPolicy: WhenUnderutilized
> ```
> 
> This basically limits consolidation to only happen during the last 15 minutes of each hour and only 1 node at a time. This is by no means a good final solution but it seems to work as expected and you can fine tune the times to prevent constant consolidation. Hope this helps anyone else looking for a temporary workaround.

With the new karpenter version 1.0.1, from this [blog post](https://aws.amazon.com/blogs/containers/announcing-karpenter-1-0/)
**Introducing consolidateAfter consolidation control for underutilized nodes**

> Karpenter prioritizes nodes to consolidate based on the least number of pods scheduled. Users with workloads that experience rapid surges in demand or interruptible jobs might have high pod churn and have asked to be able to tune how quickly Karpenter attempts to consolidate nodes to retain capacity and minimize node churn. Previously, consolidateAfter could only be used when consolidationPolicy=WhenEmpty, which is when the last pod is removed. consolidateAfter can now be used when consolidationPolicy= WhenEmptyOrUnderutilized, thus allowing users to specify in hours, minutes, or seconds how long Karpenter waits when a pod is added or removed before consolidating. If you would like the same behavior as v1beta1, then set consolidateAfter to 0 when consolidationPolicy=WhenEmptyOrUnderutilized.

**Comment by mamoit** [comment]
@joewragg It was closed on the 31st of July.
